mod/wikirate/set/type_plus_right/metric/all_values.rb

```diff
@@ -0,0 +1,231 @@
+  include Type::SearchType::HtmlFormat
+  def get_sort_icon_by_state state
+    order = state.empty? ? '' : "-#{state}"
+    %{<i class="fa fa-sort#{order}"></i>}
+  end
+
+  def toggle_sort_order order
+    order == 'asc' ? 'desc' : 'asc'
+  end
+
+  def get_sort_order sort_by, sort_order
+    if sort_by == 'company_name' 
+      [toggle_sort_order(sort_order), 'asc']
+    else
+      [get_sort_icon(''), get_sort_icon(sort_order)]
+    end
+  end
+
+  def get_sort_icon sort_by, sort_order
+    if sort_by == 'company_name' 
+      [get_sort_icon_by_state(sort_order), get_sort_icon_by_state('')]
+    else
+      [get_sort_icon_by_state(''), get_sort_icon_by_state(sort_order)]
+    end
+  end
+
+  view :card_list_header do |args|
+    sort_by = Env.params['sort_by'] || 'company_name'
+    sort_order = Env.params['sort_order'] || 'asc'
+    offset = card.get_params('offset', 0)
+    limit = card.query(search_params)[:limit]
+    company_sort_order, value_sort_order = get_sort_order sort_by, sort_order
+    company_sort_icon, value_sort_icon = get_sort_icon sort_by, sort_order
+     
+    url_template = "/#{card.cardname.url_key}?item=content&offset=#{offset}"\
+                   "&limit=#{limit}&sort_order=%s&sort_by=%s"
+    %{
+      <div class='yinyang-row column-header'>
+        <div class='company-item value-item'>
+          <a class='header metric-list-header slotter' data-remote='true' 
+            href='#{sprintf(url_template, company_sort_order, 'company_name')}'>
+            Companies #{company_sort_icon}
+          </a>
+          <a class='data metric-list-header slotter' data-remote='true' 
+            href='#{sprintf(url_template, value_sort_order, 'value')}'>
+            Values #{value_sort_icon}
+          </a>
+        </div>
+      </div>
+    }
+  end
+  # compare lenght first and then normal string comparison
+  def strcmp str1, str2
+    if (length_diff = str1.length - str2.length) == 0
+      str1 <=> str2
+    else
+      length_diff
+    end
+  end
+
+  view :card_list_item do |args|
+    c = args[:item_card]
+    item_view = inclusion_defaults(c)[:view]
+    %{
+      <div class="search-result-item item-#{ item_view }">
+        #{nest(c, size: args[:size], view: item_view)}
+      </div>
+    }
+  end
+
+  view :card_list_items do |args|
+    results =
+      search_results.map do |row|
+        c = Card["#{card.cardname.left}+#{row[0]}"]
+        render :card_list_item, args.clone.merge(item_card: c)
+      end.join "\n"
+  end
+
+  view :card_list do |args|
+    paging = _optional_render :paging, args.merge({cached_query: true})
+    if search_results.empty?
+      render_no_search_results(args)
+    else
+      results = render :card_list_items, args
+      header = render :card_list_header, args
+      %{
+        #{paging}
+        #{header}
+        <div class="search-result-list">
+          #{results}
+        </div>
+        #{paging if search_results.length > 10}
+      }
+    end
+  end
+end
```

Comment:
there is a whole lot of duplicate functionality here and in search_type.rb.  Must there be?